common/scala/src/main/scala/org/apache/openwhisk/common/Logging.scala

```diff
@@ -184,13 +185,71 @@ case class LogMarkerToken(component: String,
                           state: String,
                           subAction: Option[String] = None,
                           tags: Map[String, String] = Map.empty) {
+  private var finishToken: LogMarkerToken = _
+  private var errorToken: LogMarkerToken = _
+  private var _counter: KCounter = _
+  private var _histogram: KHistogram = _
 
   override val toString = component + "_" + action + "_" + state
-  val toStringWithSubAction =
+  val toStringWithSubAction: String =
     subAction.map(sa => component + "_" + action + "." + sa + "_" + state).getOrElse(toString)
 
-  def asFinish = copy(state = LoggingMarkers.finish)
-  def asError = copy(state = LoggingMarkers.error)
+  def asFinish: LogMarkerToken = {
+    if (finishToken == null) {
+      finishToken = copy(state = LoggingMarkers.finish)
+    }
+    finishToken
+  }
+
+  def asError: LogMarkerToken = {
+    if (errorToken == null) {
+      errorToken = copy(state = LoggingMarkers.error)
+    }
+    errorToken
+  }
```

Comment:
Is it worth caching `asFinish` or `asError` this way? Are these used multiple times in reality often?

Reply:
`asFinish` is being called from `TransactionId.finish`. So that would be frequent

Reply:
Does it make sense to just have `asFinish` and `asError` as a `lazy val` or even just a `val`?

Reply:
Thinking about it: Can all of those be `lazy val`?

Reply:
Making it val causes recursion. Lazy val I am trying to avoid as it would cause potencial concurrent access to a locked value.

Hence kept it as lazy initialized var